dirty_cat/_fuzzy_join.py

```diff
@@ -0,0 +1,176 @@
+            "precision should be either 'nearest' or",
+            f"'2dballtree', got {precision}",
+        )
+
+    if keep not in ["left", "right", "all"]:
+        raise ValueError(
+            "keep should be either 'left', 'right' or",
+            f"'all', got {keep}",
+        )
+
+    if len(suffixes)!=2:
+        raise ValueError(f"Number of suffixes specified is different than two: {suffixes}")
+    lsuffix, rsuffix = suffixes
+    if not lsuffix and not rsuffix:
+        raise ValueError(f"Tuple {suffixes} has invalid number of elements.")
+    overlap_cols = lt._info_axis.intersection(rt._info_axis)
+    if len(overlap_cols)>0:
+        for i in range(len(overlap_cols)):
+            new_name_l = overlap_cols[i] + lsuffix
+            new_name_r = overlap_cols[i] + rsuffix
+            lt.rename(columns = {overlap_cols[i]:new_name_l}, inplace = True)
+            rt.rename(columns = {overlap_cols[i]:new_name_r}, inplace = True)
+            if len(on)==2 and overlap_cols[i] in on[0]:
+                on[0] = new_name_l
+            if len(on)==2 and overlap_cols[i] in on[1]:
+                on[1] = new_name_r
+
+    if not isinstance(on, list):
+        raise ValueError(
+            f"value {on} was specified for parameter 'on', "
+            "which has invalid type, expected list of column names."
+        )
+    if len(on) == 1:
+        left_col = on[0] + lsuffix
+        right_col = on[0] + rsuffix
+    elif len(on) == 2:
+        left_col = on[0]
+        right_col = on[1]
+    else:
+        raise ValueError(
+            f"List {on} was specified for parameter 'on', "
+            "the list has invalid number of elements."
+        )
+
+    cols = list(lt.columns) + list(rt.columns)
+    joined = pd.DataFrame(lt, columns=cols)
+
+    enc = CountVectorizer(analyzer=analyzer, ngram_range=ngram_range)
+    left_enc = enc.fit_transform(lt[left_col])
+    right_enc = enc.transform(rt[right_col])
+    
+    left_enc = TfidfTransformer().fit_transform(left_enc)
+    right_enc = TfidfTransformer().fit_transform(right_enc)
+
+    # Find closest neighbor using KNN :
+    neigh = NearestNeighbors(n_neighbors=1)
+    neigh.fit(right_enc)
+    distance, neighbors = neigh.kneighbors(left_enc, return_distance=True)
+    idx_closest = np.ravel(neighbors)
+
+    if precision == 'nearest':
+        for idx in lt.index:
+            joined.loc[idx, rt.columns] = list(rt.iloc[idx_closest[idx]])
+
+    if precision == '2dball':
+        prec = []
+        for i in range(left_enc.shape[0]):
+            # Find all neighbors in a 2dball radius:
+            dist = 2 * distance[i]
+            n_neigh = NearestNeighbors(radius=dist)
+            n_neigh.fit(right_enc)
+            rng = n_neigh.radius_neighbors(left_enc[i])
+            # Distances to closest neighbors:
+            # twodball_dist = rng[0][0]
+            # Their indices:
+            twodball_pts = rng[1][0]
+            prec.append(1 / len(twodball_pts))
```

Comment:
I worry that if the shapes get a bit big, the list operations here and below are going to become a bit expensive.

Why do we need these to be a list? Are we using est_prediction? It seem to me that we are using prec, but not TP, FP, or est_precision.

Reply:
You are right, they are not useful, I removed them. But prec being a list is fine?

Reply:
Numerical operations and random access are more expensive on a list than on an array. Given that we know the length of the structure, I would rather use a preallocated array than a list that grows.

Reply:
Finally: this does not seem to me to be a "precision" in the sens that it is used in machine learning (https://en.wikipedia.org/wiki/Precision_and_recall ). We should use another variable name.

Reply:
Will change from list to array. 
I used here the term "precision" because it is called as such in the [paper I found this method being used (see Section 3.1.)](https://www.microsoft.com/en-us/research/uploads/prod/2021/03/AutoFuzzyJoin_SIGMOD21.pdf). But this is a very specific estimation of precision and recall used for fuzzy joins. I think we should dicuss further on what other, better methods can be used to estimate the precision/recall.